### PR TITLE
[storage] make epoch proof consistent with get_epoch

### PR DESCRIPTION
**Pull request by zekun000**
## Motivation

With #1948, we change the version -> epoch mapping. This PR reflects the change in get_epoch_proof too.
We return the proof from `start_epoch` to the epoch corresponding the `ledger_version`(get_epoch(ledger_version).

If start_epoch = 5, get_epoch(ledger_version) =8, the proof contains vec![5->6, 6->7, 7->8]

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

Modified tests.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

File: storage/libradb/src/ledger_store/ledger_info_test.rs [hunk 1]
```diff
@@ -48,18 +48,21 @@ proptest! {
...
             (ledger_info.epoch(), ledger_info.version())
         }).collect();
+        let start_epoch = ledger_infos_with_sigs.first().unwrap().ledger_info().epoch();
+        let latest_ledger_info = ledger_infos_with_sigs.last().unwrap().ledger_info();
+        let ledger_version = latest_ledger_info.version();
+        let proof = store.get_epoch_change_ledger_infos(start_epoch, ledger_version).unwrap();
+        let expected_proof = epoch_ledgers
+                .into_iter()
+                .filter(|li| li.ledger_info().version() < ledger_version)
+                .collect::<Vec<_>>();
```
Comment (wqfish) [inline storage/libradb/src/ledger_store/ledger_info_test.rs]: Is it okay to simply remove the last one instead of doing this filtering. Same below.

File: storage/libradb/src/ledger_store/mod.rs [hunk 2]
```diff
@@ -94,7 +94,8 @@ impl LedgerStore {
     /// Return the ledger infos reflecting epoch bumps with their least 2f+1 signatures starting
     /// from `start_epoch` to the most recent one.
     /// Note: ledger infos and signatures are available at the last version of each earlier epoch
-    /// and at the latest version of current epoch, we filter out the latter.
+    /// and at the latest version of current epoch.
+    /// We filter out the one that're equal or higher than `ledger_version`.
```
Comment (wqfish) [inline storage/libradb/src/ledger_store/mod.rs]: The old comment looks a little better.

**Comment by wqfish** [comment]
> We exclude the last ledger info corresponding to the ledger_version, so that if the ledger_version is in the middle of epoch, nothing changed, if the ledger_version has epoch change, the proof doesn't include it.

This sentence is hard to follow. Can you make it clearer.

File: storage/libradb/src/ledger_store/ledger_info_test.rs (hunk 1 above)
Comment (zekun000) [reply storage/libradb/src/ledger_store/ledger_info_test.rs]: it doesn't work. If latest_ledger_info is in the middle of one epoch, we don't pop the last one.

File: storage/libradb/src/ledger_store/mod.rs (hunk 2 above)
Comment (zekun000) [reply storage/libradb/src/ledger_store/mod.rs]: lol we're not only filtering the latter now, we're filtering the first if the last one has the same version as `ledger_version`

Comment (wqfish) [reply storage/libradb/src/ledger_store/mod.rs]: okay i understand now

File: storage/libradb/src/ledger_store/ledger_info_test.rs (hunk 1 above)
Comment (davidiw) [reply storage/libradb/src/ledger_store/ledger_info_test.rs]: it is unclear to me what this is proving aren't these two LIs equal? Maybe it should be version < version but epoch > epoch

File: storage/libradb/src/libradb_test.rs [hunk 3]
```diff
@@ -17,17 +17,20 @@ use rusty_fork::{rusty_fork_id, rusty_fork_test, rusty_fork_test_name};
...
+        })
         .cloned()
         .collect();
 
-    let (_, _, mut proof, _) = db.update_to_latest_ledger(0, Vec::new())?;
-
     // The very first validator change in the returned proof is the
     // validator change of genesis (LedgerInfo with epoch 0).
     let first_epoch_change_li = proof.ledger_info_with_sigs.remove(0);
+
```
Comment (davidiw) [inline storage/libradb/src/libradb_test.rs]: <3

really appreciate good whitespace

File: storage/libradb/src/ledger_store/mod.rs [hunk 4]
```diff
@@ -123,6 +114,17 @@ impl LedgerStore {
...
             }
+            if ledger_info_with_sigs
+                .ledger_info()
+                .next_validator_set()
+                .is_none()
+            {
+                return Err((LibraDbError::NotFound(format!(
+                    "Epoch {} change proof",
+                    ledger_info_with_sigs.ledger_info().epoch()
+                )))
```
Comment (wqfish) [inline storage/libradb/src/ledger_store/mod.rs]: This message isn't accurate. It's not `NotFound`. The error is either because the DB is corrupted, or `end_epoch` is too new so the iterator reached the latest ledger info.

File: storage/libradb/src/lib.rs [hunk 5]
```diff
@@ -620,17 +623,16 @@ impl LibraDB {
...
             .ledger_info()
             .next_validator_set()
             .is_some()
         {
-            ledger_info_with_sigs.ledger_info().epoch() + 1
-        } else {
-            ledger_info_with_sigs.ledger_info().epoch()
-        };
+            target_epoch += 1;
+        }
```
Comment (wqfish) [inline storage/libradb/src/lib.rs]: All these seem way too complex. Do I remember right that the only purpose is to get the current validator set? Can you at least add a bit comment? (I can clean up later.)

File: testsuite/tests/libratest/smoke_test.rs [hunk 6]
```diff
@@ -699,7 +699,7 @@ fn test_full_node_basic_flow() {
 
 #[test]
 fn test_e2e_reconfiguration() {
-    let (mut env, mut client_proxy_1) = setup_swarm_and_client_proxy(5, 1);
+    let (mut env, mut client_proxy_1) = setup_swarm_and_client_proxy(3, 1);
```
Comment (wqfish) [inline testsuite/tests/libratest/smoke_test.rs]: Intended?

Comment (zekun000) [reply testsuite/tests/libratest/smoke_test.rs]: Yeah, the tests would fail with 3 validators but flaky with 5 in previous broken changes. Because 3 requires everyone to make progress.

File: storage/libradb/src/ledger_store/mod.rs (hunk 4 above)
Comment (zekun000) [reply storage/libradb/src/ledger_store/mod.rs]: hmm, I didn't realize `NotFound` is for db corruption
